aws-cpp-sdk-core/source/monitoring/MonitoringManager.cpp

```diff
@@ -29,61 +29,71 @@ namespace Aws
 
         Aws::Vector<void*> OnRequestStarted(const Aws::String& serviceName, const Aws::String& requestName, const std::shared_ptr<const Aws::Http::HttpRequest>& request)
         {
-            assert(s_monitors);
             Aws::Vector<void*> contexts;
-            contexts.reserve(s_monitors->size());
-            for (const auto& interface: *s_monitors)
+            if (s_monitors)
             {
-                contexts.emplace_back(interface->OnRequestStarted(serviceName, requestName, request));
+                contexts.reserve(s_monitors->size());
+                for (const auto& interface: *s_monitors)
```

Comment:
Thanks for this change,
this is definitely a great improvement.
The only thing I can note is still a possibility for race condition:
if (s_monitors) can be true, but next line a separate thread can still destroy s_monitors.

We should merge this PR, but also follow-up on the SDK monitoring design.

Reply:
Yes that is true. I was about to add a comment for @jmklix on the issue - this PR is not a fix, as there is still a chance of the race condition to occur.

It would be great if users could _disable the MonitoringManager_,
- e.g. at compile time via a `#define`, or
- via a configuration option.

Reply:
@SergeyRyabinin - we ran into a similar situation in https://github.com/aws/aws-sdk-cpp/issues/2274. Could you please take a look to see if this is correct? To my understanding, the asynchronous threads of the `TransferManager` (used by the `S3Client`, not CRT) are the culprit, since these may complete after `main()` has returned.

I will look into a follow-up PR for this problem, since the presence of the race condition will mean that eventually a segmentation fault will occur.